File: Neocom/Neocom/NCManagedObjectObserver.m
//
//  NCManagedObjectObserver.m
//  Neocom
//
//  Created by Artem Shimanski on 21.11.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

#import "NCManagedObjectObserver.h"

@interface NCManagedObjectObserver()
@property (nonatomic, strong) NSMutableSet<NSManagedObjectID*>* objectIDs;
@property (nonatomic, copy) void(^handler)(NSSet<NSManagedObjectID*>* updated, NSSet<NSManagedObjectID*>* deleted);

@end

@implementation NCManagedObjectObserver

+ (instancetype) observerWithHandler:(void(^)(NSSet<NSManagedObjectID*>* updated, NSSet<NSManagedObjectID*>* deleted)) block {
	return [[self alloc] initWithHandler:block];
}


+ (instancetype) observerWithObjectID:(NSManagedObjectID*) objectID handler:(void(^)(NSSet<NSManagedObjectID*>* updated, NSSet<NSManagedObjectID*>* deleted)) block {
	if (!objectID)
		return nil;
	return [[self alloc] initWithObjectID:objectID handler:block];
}

- (instancetype) init {
	if (self = [super init]) {
		self.objectIDs = [NSMutableSet new];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didSave:) name:NSManagedObjectContextDidSaveNotification object:nil];
	}
	return self;
}

- (instancetype) initWithHandler:(void(^)(NSSet<NSManagedObjectID*>* updated, NSSet<NSManagedObjectID*>* deleted)) block {
	if (self = [self init]) {
		self.handler = block;
	}
	return self;
}

- (instancetype) initWithObjectID:(NSManagedObjectID*) objectID handler:(void(^)(NSSet<NSManagedObjectID*>* updated, NSSet<NSManagedObjectID*>* deleted)) block {
	if (!objectID)
		return nil;
	if (self = [self init]) {
		[self addObjectID:objectID];
		self.handler = block;
	}
	return self;
}

- (void) addObjectID:(NSManagedObjectID*) objectID {
	[self.objectIDs addObject:objectID];
}

- (void) removeObjectID:(NSManagedObjectID*) objectID {
	[self.objectIDs removeObject:objectID];
}

- (void) dealloc {
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void) didSave:(NSNotification*) notification {
	NSMutableSet* updated = [notification.userInfo[NSUpdatedObjectsKey] mutableCopy];
	NSMutableSet* deleted = [notification.userInfo[NSDeletedObjectsKey] mutableCopy];
	
	[updated intersectSet:self.objectIDs];
	[deleted intersectSet:self.objectIDs];
	if (updated.count > 0 || deleted.count > 0)
		dispatch_async(dispatch_get_main_queue(), ^{
			self.handler(updated, deleted);
		});
	/*if ([[notification.userInfo[NSUpdatedObjectsKey] valueForKey:@"objectID"] containsObject:self.objectID])
		dispatch_async(dispatch_get_main_queue(), ^{
			self.block(NCManagedObjectObserverActionUpdate);
		});
	else if ([[notification.userInfo[NSDeletedObjectsKey] valueForKey:@"objectID"] containsObject:self.objectID])
		dispatch_async(dispatch_get_main_queue(), ^{
			self.block(NCManagedObjectObserverActionDelete);
		});*/
}

@end

File: Neocom/Neocom/NCManagedObjectObserver.swift
//
//  NCManagedObjectObserver.swift
//  Neocom
//
//  Created by Artem Shimanski on 14.12.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import Foundation
import CoreData
import CloudData

class NCManagedObjectObserver {
	typealias Handler =  (_ updated: Set<NSManagedObject>?, _ deleted:Set<NSManagedObject>?) -> Void
	let handler: Handler
	var objects = Set<NSManagedObject>()
	var observer: NotificationObserver?
	
	init(managedObjects: [NSManagedObject]? = nil, handler: @escaping Handler) {
		self.handler = handler
		self.objects = Set(managedObjects ?? [])
		
		observer = NotificationCenter.default.addNotificationObserver(forName: .NSManagedObjectContextDidSave, object: nil, queue: nil /*.main*/) { [weak self] (note) in
			guard let strongSelf = self else {return}
			
			let updates: Set<NSManagedObjectID>? = {
				guard let set = note.userInfo?[NSUpdatedObjectsKey] else {return nil}
				return set as? Set<NSManagedObjectID> ?? Set((set as? Set<NSManagedObject>)?.map {$0.objectID} ?? [])
			}()
			
			let updated = strongSelf.objects.filter { (object) -> Bool in
				return updates?.contains(object.objectID) == true
			}
			
			let deletes: Set<NSManagedObjectID>? = {
				guard let set = note.userInfo?[NSDeletedObjectsKey] else {return nil}
				return set as? Set<NSManagedObjectID> ?? Set((set as? Set<NSManagedObject>)?.map {$0.objectID} ?? [])
			}()

			let deleted = strongSelf.objects.filter { (object) -> Bool in
				return deletes?.contains(object.objectID) == true
			}

//			let updated = Set((note.userInfo?[NSUpdatedObjectsKey] as? Set<NSObject>)?.flatMap{ ($0 as? NSManagedObject)?.objectID ?? $0 as? NSManagedObjectID} ?? []).intersection(objects)
//			let deleted = Set((note.userInfo?[NSDeletedObjectsKey] as? Set<NSObject>)?.flatMap{ ($0 as? NSManagedObject)?.objectID ?? $0 as? NSManagedObjectID} ?? []).intersection(objects)
			
			//let updated = (note.userInfo?[NSUpdatedObjectsKey] as? Set<NSManagedObjectID>)?.intersection(strongSelf.objectIDs)
			//let deleted = (note.userInfo?[NSDeletedObjectsKey] as? Set<NSManagedObjectID>)?.intersection(strongSelf.objectIDs)
			if !updated.isEmpty || !deleted.isEmpty {
				DispatchQueue.main.async {
					strongSelf.handler(Set(updated), Set(deleted))
				}
			}
		}
	}
	
	convenience init(managedObject: NSManagedObject, handler: @escaping Handler) {
		self.init(managedObjects: [managedObject], handler: handler)
	}
	
	deinit {
		NotificationCenter.default.removeObserver(observer!)
	}

	func add(managedObject: NSManagedObject) {
		objects.insert(managedObject)
	}
	
	func remove(managedObject: NSManagedObject) {
		objects.remove(managedObject)
	}
	
}
